Panic server on watch errors during etcd test

enj opened the pull request:
**What this PR does / why we need it**:

This change makes it so that errors during watch decoding panic the server during the etcd storage test.  This allows us to catch coder errors related to storing incompatible types at the same location in etcd.

For example, https://github.com/kubernetes/kubernetes/pull/39164#discussion_r116619172 would have failed like so (instead of silently recording an error to the master log and passing the test):

```
I0518 14:39:32.413038    4654 storage_rbac.go:257] created rolebinding.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-public
panic: v1beta1.NetworkPolicy is not suitable for converting to {{"networking.k8s.io" "__internal"} [{"networking.k8s.io" ""} {"networking.k8s.io" ""}]}

goroutine 799 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.decodeObj(0x71146c0, 0xc420eb9d80, 0x7129340, 0x733cdf8, 0xc423708360, 0x118, 0x120, 0x86, 0x0, 0x0, ...)
	/home/mkhan/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:382 +0x115
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).prepareObjs(0xc420958f60, 0xc422ee9ef0, 0xc42003c600, 0x8a3b9b, 0x70f36c0, 0xc4209585a0, 0x4583f0, 0xc4215d0f58)
	/home/mkhan/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:353 +0x2cd
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).transform(0xc420958f60, 0xc422ee9ef0, 0x0)
	/home/mkhan/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:248 +0x4d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).processEvent(0xc420958f60, 0xc420132010)
	/home/mkhan/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:213 +0x122
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).run
	/home/mkhan/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:116 +0xe9
exit status 2
FAIL	k8s.io/kubernetes/test/integration/etcd	13.065s
Makefile:164: recipe for target 'test' failed
```

@liggitt PTAL

Signed-off-by: Monis Khan <mkhan@redhat.com>

**Release note**:

```
NONE
```

liggitt commented:
@k8s-bot ok to test

timothysc commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -376,6 +379,9 @@ func (wc *watchChan) prepareObjs(e *event) (curObj runtime.Object, oldObj runtim
 	obj, err := runtime.Decode(codec, []byte(data))
 	if err != nil {
+		if fatalOnDecodeError {
+			panic(err) // we are running in a test environment and thus an error here is due to a coder mistake
```
Umm this is a too agressive imo, why panic?  Also under what conditions are you wanting to do this?  

/cc @wojtek-t

wojtek-t replied:
I'm really against the panic. If we have corrupted data in etcd, apiserver won't be able to start with panic. With error, we will simply be returning errors.

liggitt replied:
We want to enable this in test mode so we don't silently fail on colocated resources encoding incorrectly (as we do today). This is like the cache mutation detector, which only runs in test mode, and panics if you mutate an object in a shared informer cache

timothysc replied:
Error log detection seems like a reasonable approach here.

liggitt replied:
that seems weak... I'd be worried about pattern match drift

timothysc replied:
> I'd be worried about pattern match drift 

Then use a const for the message.

wojtek-t replied:
OK - I got the motivation now and that sounds fine to me.
In that case, I vote for consistency in all those mechanism and suggest using some env var to enable/disable it. For the mutation detector we have this:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/mutation_detector.go#L34

enj replied:
@wojtek-t I changed it to match the cache mutation detector.

timothysc commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -384,3 +390,8 @@ func decodeObj(codec runtime.Codec, versioner storage.Versioner, data []byte, re
 }
+
+// TestingPanicServerOnWatchDecodeError is used during testing to make watch decoding errors fatal
+func TestingPanicServerOnWatchDecodeError() {
```
Test code should not be part of the main library.

liggitt replied:
suggestions for flipping the behavior for test mode are welcome

fejta commented:
/unassign
/assign @liggitt

enj committed:
Panic server on watch errors in test environment

This change makes it so that errors during watch decoding panic the
server if it is in a test environment.  This allows us to catch coder
errors related to storing incompatible types at the same location in
etcd.

Signed-off-by: Monis Khan <mkhan@redhat.com>

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:
```diff
@@ -224,6 +224,12 @@ func TestWatchFromNoneZero(t *testing.T) {
 }
 
 func TestWatchError(t *testing.T) {
+	// we need to check for watch errors so do not panic the server
```
This can cause races if tests run in parallel, right?

enj replied:
@wojtek-t yeah, though I do not know of a better way that does not require using locks.

wojtek-t replied:
@deads2k - do you think we should care about this race?

deads2k replied:
> @deads2k - do you think we should care about this race?

I think we can restructure the test to catch the expected behavior.  If we're set to panic, use a defer/recover construct.  If we're set to not-panic, expect the error.

wojtek-t replied:
Yeah that's a good point.
+1 for restructuring this test.

enj replied:
@deads2k @wojtek-t so the `panic` from the decode error happens in a different go routine and thus I do not think the test can directly `recover` from it.  I added some code that uses a `defer` in the non-test code, but it all seems a bit hacky to me.

enj commented:
@wojtek-t comments?

wojtek-t commented on hack/local-up-cluster.sh:
```diff
@@ -658,11 +662,11 @@ function start_kubelet {
       KUBELET_PID=$!
       # Quick check that kubelet is running.
-      if ps -p $KUBELET_PID > /dev/null ; then 
+      if ps -p $KUBELET_PID > /dev/null ; then
```
Thanks for those fixes !

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -373,9 +394,14 @@ func (wc *watchChan) prepareObjs(e *event) (curObj runtime.Object, oldObj runtim
 	obj, err := runtime.Decode(codec, []byte(data))
 	if err != nil {
+		if fatalOnDecodeError {
+			defer testingDeferOnDecodeError() // catch watch decode error iff we caused it on purpose during a unit test
```
Can you please split the lines a bit so that they aren't that long?

wojtek-t reviewed (commented):
Just very minor comment - other than that LGTM.

enj commented:
@wojtek-t all comments addressed.

wojtek-t commented:
/lgtm
/approve
